modules/social_features/social_comment/tests/src/Kernel/CommentViewAccessTest.php

```diff
@@ -0,0 +1,181 @@
+    // Create an unpublished comment on a node.
+    $user = $this->createUser([], ['access comments']);
+    $this->setCurrentUser($user);
+    $this->createComment($this->node, $user, ['status' => 1]);
+
+    $this->assertNotEmpty($this->getCommentIds($this->node, $user));
+  }
+
+  /**
+   * Test that a user can not view another person's unpublished comment.
+   */
+  public function testUserCanNotViewOtherUnpublishedComment() {
+    // Create an unpublished comment on a node.
+    $first_user = $this->createUser([], ['access comments']);
+    $this->createComment($this->node, $first_user, ['status' => 0]);
+
+    // Create another user to try and view the comment.
+    $second_user = $this->createUser([], ['access comments']);
+    $this->setCurrentUser($second_user);
+
+    $this->assertEmpty($this->getCommentIds($this->node, $first_user));
+  }
+
+  /**
+   * Test that a user can view another person's published comment.
+   */
+  public function testUserCanViewOtherPublishedComment() {
+    // Create an unpublished comment on a node.
+    $first_user = $this->createUser([], ['access comments']);
+    $this->createComment($this->node, $first_user, ['status' => 1]);
+
+    // Create another user to try and view the comment.
+    $second_user = $this->createUser([], ['access comments']);
+    $this->setCurrentUser($second_user);
+
+    $this->assertNotEmpty($this->getCommentIds($this->node, $first_user));
+  }
+
+  /**
+   * Create the comment entity.
+   *
+   * @param \Drupal\Core\Entity\EntityInterface $entity
+   *   The entity the comment is made on.
+   * @param \Drupal\Core\Session\AccountInterface|null $user
+   *   An optional user to create the comment as.
+   * @param mixed[] $values
+   *   An optional array of values to pass to Comment::create.
+   *
+   * @throws \Drupal\Core\Entity\EntityStorageException
+   */
+  private function createComment(EntityInterface $entity, ?AccountInterface $user = NULL, array $values = []): void {
+    if ($user !== NULL) {
+      $values += ['uid' => $user->id()];
+    }
+
+    $this->storage->create(
+      $values +
+      [
+        'entity_id' => $entity->id(),
+        'entity_type' => $entity->getEntityTypeId(),
+        'comment_type' => 'comment',
+        'field_name' => 'comments',
+      ]
+    )->save();
+  }
+
+  /**
+   * Get a list of comment IDs whose user have access.
+   *
+   * @param \Drupal\Core\Entity\EntityInterface $entity
+   *   The entity the comment is made on.
+   * @param \Drupal\Core\Session\AccountInterface $user
+   *   The user to create the comment as.
+   *
+   * @return array
+   *   An array of comment IDs.
+   */
+  private function getCommentIds(EntityInterface $entity, AccountInterface $user): array {
+    return $this->storage
+      ->getQuery()
+      ->currentRevision()
+      ->accessCheck()
+      ->condition('entity_id', $entity->id())
+      ->condition('comment_type', 'comment')
+      ->condition('uid', $user->id())
+      ->execute();
```

Comment:
I don't think this properly tests the access check. By adding the `condition('uid', $user->id())` we only load the entities owned by the passed in user, rather than the ones they have access to.

I also think it's a good idea to just move this out of the function and into the tests so it's clearer what is being tested (the accessCheck). Then you could even perform the same query without the access check to show that that works as well, e.g.

```php
// [ create two comments of which the current user can see 1 ]
$all_comments = $this->storage
      ->getQuery()
      ->accessCheck(FALSE)
      ->condition('entity_id', $entity->id())
      ->condition('comment_type', 'comment')
      ->execute();
self::assertCount(2, $all_comments);
$visible_comments = $this->storage
      ->getQuery()
      ->accessCheck(TRUE)
      ->condition('entity_id', $entity->id())
      ->condition('comment_type', 'comment')
      ->execute();
self::assertCount(1, $visible_comments);
```